app/models/govuk_publishing_components/audit_components.rb

```diff
@@ -215,5 +225,50 @@ def get_asset_link(a_thing, component)
 
       link
     end
+
+    def get_helper_usage(components)
+      helpers = [
+        {
+          name: "Brand helper",
+          link: "lib/govuk_publishing_components/app_helpers/brand_helper.rb",
+          match: /(GovukPublishingComponents::AppHelpers::BrandHelper.new)/,
+          used_by: [],
+        },
+        {
+          name: "Component wrapper helper",
+          link: "lib/govuk_publishing_components/presenters/component_wrapper_helper.rb",
+          match: /(GovukPublishingComponents::Presenters::ComponentWrapperHelper.new)/,
+          used_by: [],
+        },
+        {
+          name: "Heading helper",
+          link: "lib/govuk_publishing_components/presenters/heading_helper.rb",
+          match: /(GovukPublishingComponents::Presenters::HeadingHelper.new)/,
+          used_by: [],
+        },
+        {
+          name: "Shared helper",
+          link: "lib/govuk_publishing_components/presenters/shared_helper.rb",
+          match: /(GovukPublishingComponents::Presenters::SharedHelper.new)/,
+          used_by: [],
+        },
+      ]
+
+      components.each do |component|
+        file = component[:template_file]
+        next unless File.file?(file)
+
+        helpers.each do |helper|
+          next if File.foreach(file).grep(helper[:match]).empty?
+
+          helper[:used_by] << {
+            name: component[:name],
+            link: component[:template_file].split("/components/")[1],
+          }
+        end
+      end
```

Comment:
If the files are small, something like this maybe more efficient...

```ruby
components.each do |component|
  next unless File.file?(component[:template_file])

  # This reads the entire file into memory, so not good for large files. Also, you need to remember to `.close` it to avoid leaks
  file = File.open(component[:template_file])
  helpers.each do |helper|
    # I've reversed the logic a little, think it's easier to read given the `next` above
    next unless file.grep(helper[:match]).present?  

    helper[:used_by] << {
      name: component[:name],
      link: component[:template_file].split("/components/")[1],
    }
  end
  file.close
end
```
Otherwise, I think what you've done is fine.

But, the performance will also depend on which is the shorter list `components` or `helpers`. As a general rule, whichever is predictably the shorter put that on the outside loop.

Reply:
Thanks. I deliberately avoided `File.open` on the basis that some are large files, because I read that `File.foreach(file).grep` reads the file line by line, so doesn't load it all into memory at once. Most of the time the lines we're looking for will be in the first 10 lines or so of the template, so that seemed like an efficient way to go.